lxd/db/query/dump.go

```diff
@@ -10,21 +10,37 @@ import (
 	"time"
 )
 
-// Dump returns a SQL text dump of all rows across all tables, similar to
-// sqlite3's dump feature
-func Dump(ctx context.Context, tx *sql.Tx, schema string, schemaOnly bool) (string, error) {
+// Dump returns a SQL text dump of all rows across all tables, similar to  sqlite3's dump feature.
+// The "extraTables" argument can be populated with a map of table name to table schema. If a given table name is known
+// at the package level, the schema value can be left blank.
+func Dump(ctx context.Context, tx *sql.Tx, schema string, schemaOnly bool, extraTables map[string]string) (string, error) {
```

Comment:
I think now would be a good time to refactor this function more to remove the special case of injecting the existing schema into this function in order for it to ascertain which tables to dump, rather than adding more special cases by way of the extraTables argument.

Do you think instead we can just have a `tables ...string` argument and then call `dumpParseSchema(schema)` from outside of this function to get the table names to retrieve from the database.

The blocker I can see is returning the schema itself. It seems odd to me that this function doesn't return schema from the DB but instead returns it from the current definition (which may or may not be the one in the actual database - depending on if the schema patches have been applied yet).

I wonder if we can retrieve the schema for the table from the database directly?

Reply:
At the very least we could move the call to `dumpParseSchema(schema)` out of this function (and into the caller), and then have a single argument called `tables map[string]string` which would inform which tables need to be dumped, and we would expect the caller to add "schema" to it as well using `dumpSchemaTable`.

Reply:
We should be able to use `SELECT sql FROM sqlite_master WHERE name='<table>'` to get the actual schema in the DB btw.

Reply:
I agree, this is a good time to do the refactor. I was deliberating on this myself and took the easier option ;)

I'll attempt to get all of the required info from the DB itself without relying on the current definition. If not I'll move the `dumpParseSchema` call to the callers as you suggest.

Reply:
Thanks! That would be good as I myself have found strange issues with this command in the past where the schema returned does not match what the actual DB itself has (because the schema patches got out of sync with the DB).

Reply:
There can also be differences in how the schema is returned vs how it was requested to be created so should be better to "get it from the horses mouth".

Reply:
@tomponline I've implemented this now getting everything direct from the database :smile: . Can you take another look please?